File: DTSparkCore/src/main/java/com/kevin/java/secondsort/SecondarySort.java
package com.kevin.java.secondsort;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;

/**
 * @author kevin
 * @version 1.0
 * @description     二次排序
 * @createDate 2019/1/2
 */
public class SecondarySort {

    public static void main(String[] args) {

        SparkConf conf = new SparkConf().setAppName("SecondarySort").setMaster("local");
        JavaSparkContext sc = new JavaSparkContext(conf);

        String file = "DTSparkCore\\src\\main\\resources\\secondSort.txt";
        JavaRDD<String> lines = sc.textFile(file);

        // 将数据封装成SecondSortKey类型
        JavaPairRDD<SecondSortKey, String> pairRDD = lines.mapToPair(new PairFunction<String, SecondSortKey, String>() {
            @Override
            public Tuple2<SecondSortKey, String> call(String line) throws Exception {
                String[] splited = line.split(" ");
                int first = Integer.valueOf(splited[0]);
                int second = Integer.valueOf(splited[1]);
                SecondSortKey secondSortKey = new SecondSortKey(first, second);
                return new Tuple2<SecondSortKey, String>(secondSortKey, line);
            }
        });
        // 重写排序方法，在使用排序的时候应用其自定义的排序方法
        pairRDD.sortByKey(false).foreach(new VoidFunction<Tuple2<SecondSortKey, String>>() {
            @Override
            public void call(Tuple2<SecondSortKey, String> tuple2) throws Exception {
                System.out.println(tuple2._2);
            }
        });

        sc.close();

    }
}

File: dt-spark-core/src/main/scala/com/kevin/scala/secondsort/SecondarySort.scala
package com.kevin.scala.secondsort

import org.apache.spark.{SparkConf, SparkContext}

/**
  * 二次排序
  */
object SecondarySort {

  def main(args: Array[String]): Unit = {
    // 1.创建SparkConf
    val conf = new SparkConf().setAppName("SecondarySort").setMaster("local")
    // 2.创建SparkContext
    val sc = new SparkContext(conf)
    val file = "DTSparkCore\\src\\main\\resources\\secondSort.txt"
    // 3.textFile读取文件数据,map将数据封装成SecondSortKey类型,sortByKey重写排序方法,使用自定义的排序方法,foreach遍历输出
    sc.textFile(file).map(line => {
      val first = line.split(" ")(0).toInt
      val second = line.split(" ")(1).toInt
      (new SecondSortKey(first,second),line)
    }).sortByKey(false).foreach(value => println(value._2))
    // 4.关闭sc
    sc.stop()

  }

}
